addons/misra.py

```diff
@@ -256,14 +256,105 @@ def getForLoopExpressions(forToken):
+            ret.extend(findCounterTokens(cond.astOperand2))
+        elif cond.astOperand1.isOp and cond.astOperand2.isNumber:
+            ret.extend(findCounterTokens(cond.astOperand1))
+        elif cond.astOperand2.isOp and cond.astOperand1.isNumber:
+            ret.extend(findCounterTokens(cond.astOperand2))
+
+    if cond.isComparisonOp:
+        if cond.astOperand1.isName and cond.astOperand2.isNumber:
+            ret.append(cond.astOperand1)
+        elif cond.astOperand2.isName and cond.astOperand1.isNumber:
+            ret.append(cond.astOperand2)
+        elif cond.astOperand2.isOp and cond.astOperand1.isOp:
+            ret.extend(findCounterTokens(cond.astOperand1))
+            ret.extend(findCounterTokens(cond.astOperand2))
+        elif cond.astOperand1.isOp and cond.astOperand2.isNumber:
+            ret.extend(findCounterTokens(cond.astOperand1))
+        elif cond.astOperand2.isOp and cond.astOperand1.isNumber:
+            ret.extend(findCounterTokens(cond.astOperand2))
+
+    return ret
+
+
+def getFloatComparision(floatCompOperator):
+    while not floatCompOperator.isComparisonOp:
+        floatCompOperator = floatCompOperator.astParent
+    expr = hasFloatComparison(floatCompOperator)
+    if expr:
+        return True
+    return False
+
+
+def getWhileLoopExpressions(whileToken):
+    if not simpleMatch(whileToken, 'while ('):
+        return None
+    lpar = whileToken.next
+    if not lpar.astOperand2.astOperand1:
+        return None
+    rpar = lpar.link
+    counter = findCounterTokens(lpar.astOperand2)
+    if len(counter) == 0:
+        return None
+    if simpleMatch(rpar, ') {'):
+        token = rpar.next
+        while (token != rpar.next.link):
+            token = token.next
+            for counter_str in counter:
+                if token.isAssignmentOp and token.astOperand1.str == counter_str.str:
+                    return getFloatComparision(counter_str.astParent)
+                elif token.str == counter_str.str and token.astParent and token.astParent.str in {'++', '--'}:
+                    return getFloatComparision(counter_str.astParent)
+    elif simpleMatch(whileToken.previous, '} while') and simpleMatch(whileToken.previous.link.previous, 'do {'):
+        token = whileToken.previous
+        while (token != whileToken.previous.link):
+            token = token.previous
+            for counter_str in counter:
+                if token.isAssignmentOp and token.astOperand1.str == counter_str.str:
+                    return getFloatComparision(counter_str.astParent)
+                elif token.str == counter_str.str and token.astParent and token.astParent.str in {'++', '--'}:
+                    return getFloatComparision(counter_str.astParent)
+    return None
+
+
 def hasFloatComparison(expr):
     if not expr:
         return False
     if expr.isLogicalOp:
         return hasFloatComparison(expr.astOperand1) or hasFloatComparison(expr.astOperand2)
+    if expr.isArithmeticalOp:
```

Comment:
it seems to me that with this code:

    a = b + 5.0;

you will say that there is a float comparison. I do not agree we should return True for this code.

This function used to check if there was a comparison.. and if that was seen it was checked if one of the operands was a float type.

Reply:
I included this so that  while((10.0f*c) > (a*b)) can return True.

Reply:
As far as I see the old code should return True for that. Maybe I need to look into this. Did the cppcheckdata.astIsFloat() return False for the expression `10.0f * c`? Does it matter if c is declared?